airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java

```diff
@@ -347,6 +348,14 @@ public void createUser(User user, UserCredential credential) {
      */
     @CacheEvict
     public void deleteUser(String username) {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User currentUser = getUserByName(authentication.getName());
+        if (!currentUser.isAdminRole()) {
+            throw new org.springframework.security.access.AccessDeniedException("Only administrators can manage users");
+        }
+        if (currentUser.getUsername().equals(username)) {
+            throw new IllegalArgumentException("Can't delete yourself");
+        }
```

Comment:
> prevent deletion only when there's a single user with admin privileges left

That was my first implementation, TBH.

> the current setup allows a user with admin rights to delete their own account

That's not absolutely true. Currently admin can delete yourself via Subsonic REST API, but not via Web UI: there is just no `Delete this user` checkbox.

Because of above, I've changed my mind: the current administrator
> will never be deleted, which essentially guarantees that there will always be at least one user with admin privileges

> restriction that prevents users with admin privileges from deleting their own account, which I feel is a step back from the existing functionality

1. Not step back, because it's current behavior in UI.
2. From my point of view it's right. Otherwise it looks like shooting yourself in the foot.
3. If I login by `other_adm` and delete myself via REST `http://airsonic.local:8080/rest/deleteUser?username=other_adm`, then I get [error in UI](https://github.com/kagemomiji/airsonic-advanced/files/12327622/error-after-user-deletion.txt)

Reply:
@zc-devs 
Thank you for sharing your perspective.
I acknowledge and agree with your point that this isn't really a step back given the current behavior in the UI.

Pulling the currently authenticated user directly from the Context in the Service layer feels a bit intertwined. Perhaps, we could consider an alternative?

How about fetching the authenticated user in the Controller layer and then relay that data to the Service layer? This way, the Service method could simply take in the user data and perform its comparison with the username. I believe this approach might be more testable and provide greater flexibility for potential future enhancements. I'd love to hear your thoughts on this!

https://github.com/kagemomiji/airsonic-advanced/blob/c592afa949fe4225b8778600b01c28502acc05ed/airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java#L352-L356

Reply:
Let's try that ([85ea6cd](https://github.com/kagemomiji/airsonic-advanced/pull/263/commits/85ea6cddc1ca08efa791e6feac45f042e868b31f))